go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -184,18 +341,97 @@ func (npw *nodePortWatcher) SyncServices(services []interface{}) {
 				serviceInterface)
 			continue
 		}
-		npw.updateServiceFlowCache(service, true)
+		npw.updateServiceFlowCache(service, true, false)
 	}
 
 	npw.ofm.requestFlowSync()
-	syncSharedGatewayIptRules(services)
+	syncSharedGatewayIptRules(services, false, npw.nodeIPs)
+}
+
+// Add new Local Traffic flows if endpoints are local
+func (npw *nodePortWatcher) AddEndpoints(ep *kapi.Endpoints) {
+	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
+	// Do nothing if service dosn't exist yet
+	if _, exists := npw.services[name]; exists {
+		// If ETP isn't Local Do nothing
+		if util.ServiceExternalTrafficPolicyLocal(npw.services[name]) {
+			// Only update flows if node local ep state changed from false -> true
+			// Don't care about true -> true or false -> false since correct flows should already be in place
+			if countHostNeworkedEndpoints(ep, npw.nodeName) > 0 {
+				klog.V(5).Infof("Service %s has ETP=local and local Endpoints, Updating flow Cache", ep.Name)
+				// Update Service Cache with special flows
+				npw.updateServiceFlowCache(npw.services[name], true, true)
+				npw.ofm.requestFlowSync()
+				//Service was already created so now delete old rules and add correct IPtables rules
+				delSharedGatewayIptRules(npw.services[name], false, npw.nodeIPs)
+				addSharedGatewayIptRules(npw.services[name], true, npw.nodeIPs)
+			}
+		}
+	}
+}
+
+func (npw *nodePortWatcher) DeleteEndpoints(ep *kapi.Endpoints) {
+	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
+	if _, exists := npw.services[name]; exists {
+		if util.ServiceExternalTrafficPolicyLocal(npw.services[name]) {
+			if countHostNeworkedEndpoints(ep, npw.nodeName) == 0 {
+				klog.V(5).Infof("Service %s has ETP=local and no local Endpoints, Updating flow Cache", ep.Name)
+				npw.updateServiceFlowCache(npw.services[name], false, true)
+				npw.ofm.requestFlowSync()
+				delSharedGatewayIptRules(npw.services[name], true, npw.nodeIPs)
+			}
+		}
+	}
+	// We add this only when a Service event is received, delete only when endpoint delete event is received
+	delete(npw.services, name)
+}
+
+func (npw *nodePortWatcher) UpdateEndpoints(old, new *kapi.Endpoints) {
```

Comment:
suggestion: make `AddEndpoints` just call `UpdateEndpoints(nil, new)`

Reply:
Not sold on this yet, the logic is annoying since we're managing iptables rules and flows from one event, and we need to make sure on update that things are properly cleaned up